include/picongpu/plugins/openPMD/openPMDWriter.hpp

```diff
@@ -0,0 +1,1446 @@
+#include <list>
+#include <pthread.h>
+#include <sstream>
+#include <string>
+#include <vector>
+
+
+namespace picongpu
+{
+namespace openPMD
+{
+    using namespace pmacc;
+
+
+    namespace po = boost::program_options;
+
+    template< unsigned DIM >
+    ::openPMD::RecordComponent &
+    ThreadParams::initDataset(
+        ::openPMD::RecordComponent & recordComponent,
+        ::openPMD::Datatype datatype,
+        pmacc::math::UInt64< DIM > const & globalDimensions,
+        bool compression,
+        std::string const & compressionMethod )
+    {
+        std::vector< uint64_t > v = asStandardVector( globalDimensions );
+        ::openPMD::Dataset dataset{ datatype, std::move( v ) };
+        if( compression && compressionMethod != "none" )
+        {
+            dataset.compression = compressionMethod;
+        }
+        recordComponent.resetDataset( std::move( dataset ) );
+        return recordComponent;
+    }
+
+
+    template< typename Vec, typename Ret >
+    Ret
+    asStandardVector( Vec const & v )
+    {
+        using Vec_T = typename std::remove_reference< Vec >::type;
+        constexpr auto DIM = Vec_T::dim;
+        Ret res( DIM );
+        for( unsigned i = 0; i < DIM; ++i )
+        {
+            res[ DIM - i - 1 ] = v[ i ];
+        }
+        return res;
+    }
+
+    ::openPMD::Series &
+    ThreadParams::openSeries( ::openPMD::AccessType at )
+    {
+        if( !openPMDSeries )
+        {
+            std::string fullName = fileName + fileInfix + "." + fileExtension;
+            log< picLog::INPUT_OUTPUT >( "openPMD: open file: %1%" ) % fullName;
+            openPMDSeries = std::unique_ptr<::openPMD::Series >(
+                new ::openPMD::Series( fullName, at, communicator, jsonConfig ) );
+            if( at == ::openPMD::AccessType::CREATE )
+            {
+                openPMDSeries->setMeshesPath( MESHES_PATH );
+            }
+            log< picLog::INPUT_OUTPUT >(
+                "openPMD: successfully opened file: %1%" ) %
+                fullName;
+            return *openPMDSeries;
+        }
+        else
+        {
+            throw std::runtime_error(
+                "openPMD: Tried opening a Series while old Series was still "
+                "active" );
+        }
+    }
+
+    void
+    ThreadParams::closeSeries()
+    {
+        if( openPMDSeries )
+        {
+            log< picLog::INPUT_OUTPUT >( "openPMD: close file: %1%" ) %
+                fileName;
+            openPMDSeries.reset();
+            MPI_Barrier( this->communicator );
+            log< picLog::INPUT_OUTPUT >(
+                "openPMD: successfully closed file: %1%" ) %
+                fileName;
+        }
+        else
+        {
+            throw std::runtime_error(
+                "openPMD: Tried closing a Series that was not active" );
+        }
+    }
+
+
+
+    bool
+    ThreadParams::isADIOS1()
+    {
+#if openPMD_HAVE_ADIOS1 && !openPMD_HAVE_ADIOS2
+        return this->fileExtension == "bp";
```

Comment:
Is it still possible to write explicit ADIOS1 data when OpenPMD is compiled for Adios1 and Adios2?

Reply:
Ah yes, in the meantime the environment variable `openPMD_BP_BACKEND` has been added to the openPMD API, so I should check for that too. I've added a todo to remove this once the corresponding [openPMD API bug](https://github.com/openPMD/openPMD-api/issues/520) is resolved.